lib/Fmt_ast.ml

```diff
@@ -3433,6 +3438,28 @@ and fmt_module_type c ?(rec_ = false) ({ast= mty; _} as xmty) =
           fmt_or_k (Option.is_none blk.pro)
             (fits_breaks " " ~hint:(1, 2) "")
             blk.psp }
+  | Pmty_gen (gen_loc, mt) ->
+      let blk = fmt_module_type c (sub_mty ~ctx mt) in
+      let pre =
+        if V.at_least c.conf Ocaml_version.Releases.v5_1_0 then noop
+        else
+          str "functor"
+          $ fmt_attributes c ~pre:Blank pmty_attributes
+          $ fmt "@;<1 2>"
```

Comment:
Now that there's a new constructor, I think this cause more problems than it solves.

If the syntax was present in the source file, I think it can be safely outputted.

Reply:
Yes but the ocaml-version option allows the user to parse a file with 5.1.0 and produce an output for 4.08 for example

Reply:
Is this the intended purpose of this option ? I've never taken this into account and I don't think we can ensure this is correctly handled.
This is in contradiction with the direction we are taking regarding parsing.

If this is something we want to do in this case (I don't think so) it'll be better handled as a transformation pass on the AST rather than adding duplicated/inconsistent formatting code.

Reply:
That's useful if people are not all using the same version of ocaml on the same project. I'm not saying it's the best use of ocamlformat, but that's what we have done for new syntaxes in the past (I have punning in mind, but maybe more). I think @hhugo emitted the idea if I'm not mistaken.

Reply:
We didn't have a concrete AST at this time and we needed to avoid sugaring existing code into new syntaxes. 

I don't agree with this rewrite being done in the formatting function and I'd prefer to move that into a transformation pass (similarly to how punning is handled now).

I suggest to move this conversation in a new PR/issue and merge this PR without the rewriting feature.

Reply:
I pushed a commit to remove the automatic syntax downgrade and I'll merge. Happy to discuss the proposal in a new PR.

This is necessary due to the concrete AST: we can safely output the new syntax because it was also present in the input. The version of ocaml required to parse the output file is identical to the input file.

That's also the direction we are taking with the more concrete AST.